### ISSUE TITLE
[Component amendment]: Badge title / aria-label and allowed content

### ISSUE THREAD

**Issue by dgtlntv** [canonical/vanilla-framework#5619]
While @jmuzina was working on the implementation of the Badge in the new repo two questions came up.
1. Wether we should allow other numerical information like eg. less than (<20) or ranges (10-20) or if the badge should exclusively be limited to only showing quantities (eg 5, 999+, 2T+).
2. What a11y approach we should take for abbreviated large quantities. aria-labels should apparently only be used when a specific aria-role is used. Should we ignore the best practice and always set the aria-label in this case? Additionally should we set a title property (ie. browser native tooltip) with the full quantity? And lastly should the quantity be spelled out in words or be formatted as a number?

**Comment by jmuzina** [comment]
Adding some more more details:

Currently, Vanilla uses `<span>` with aria-label to spell out badge text. However, in Pragma, our Biome linter threw up a warning that using `aria-label` with `span` without adding accessible aria roles is illegal, and browsers will not announce the label text. Here's an [article](https://benmyers.dev/blog/dont-use-aria-label-on-static-text-elements/) with more details on this.

So far, in my [draft implementation on Pragma](https://github.com/canonical/pragma/pull/302), I have been only adding the aria label if `role="status"` is used. This then causes the badge to be spelled out:

<img width="183" height="90" alt="Image" src="https://github.com/user-attachments/assets/5a978ddf-d87f-42bd-9bf8-f9a30e987968" />
<img width="295" height="90" alt="Image" src="https://github.com/user-attachments/assets/b3f091ad-3b28-4ffd-8ddf-206d683efc72" />
<img width="295" height="90" alt="Image" src="https://github.com/user-attachments/assets/c78a8f59-0249-4939-a4c5-f8b2de088d96" />

([story link](https://67be30a0b85c4e44b6744158-fpgeqlsyxz.chromatic.com/?path=/story/badge--with-role&args=value:10001;overflowStrategy:compact) if you want to play around with it more)

It would be helpful if we can define:
1. How to make the badge text accessible
2. Whether to (and how to) include the value in accessible text. What language should be used - when should "approximately" be used, how should we indicate that the value is more than can fit into the badge, etc.

**Comment by juanruitina** [comment]
Some thoughts about accessible labeling.

TL;DR:

- I think screen readers handle numerals well, no need to spell them out.
- I find the current titles perhaps too verbose, something like "more than 10000 items" will be good enough.
- I'd think there's as much a need to explain "k" and "m" abbreviations to sighted users as to non-sighted ones, but not a good solution that works for all. Using a combination of `aria-hidden="true"` and the `.u-off-screen` class might be a cleaner way to at least ensure a passing, safe experience for screen-reader users.

---
 
From the article you link to:

> On a similar token, sometimes aria-label is used to provide [verbose, redundant, potentially off-base control hints](https://adrianroselli.com/2019/10/stop-giving-control-hints-to-screen-readers.html), which we could opt to rip out.

I wonder to what extent we need such a verbose label. I don't think "10k+" is more informative visually than spoken. How much affordance is lost without the visuals? Maybe that it is a count of items. Generally off-screen content is safer with screen readers than `aria-label` and `title` so perhaps this would be good enough:

```html
<span class="ds badge">
  999+<span class="u-off-screen"> items</span>
</span>
```

```html
<span class="ds badge">
  10k+<span class="u-off-screen"> items</span>
</span>
```

```html
<span class="ds badge">
  <span class="u-off-screen">approximately </span>1k<span class="u-off-screen"> items</span>
</span>
```

Applying the same logic to abbreviated large quantities: we may decide that "k" or "m" need explanation, but whatever we decide applies to both sighted and non-sighted users. I see however that's difficult to achieve.

If we decide to explain it, the most standard way would be to use `abbr` with a `title`, but yep, there we have a problem. Although markup would be formally correct, `abbr` is not a super common pattern, it's effectively unsupported on mobile, and also [problematic](https://www.tempertemper.net/blog/abbreviations-can-be-problematic) accessibility-wise.

So we could rely on `aria-hidden` instead ([more info](https://www.accessibility-developer-guide.com/examples/sensible-aria-usage/hidden/)) so at least it's explained to screen reader users:

```html
<span class="ds badge">
  <span aria-hidden="true">999+</span>
  <span class="u-off-screen">more than 999 items</span>
</span>
```

```html
<span class="ds badge">
  <span aria-hidden="true">10k+</span>
  <span class="u-off-screen">more than 10000 items</span>
</span>
```

(I think "10k+" should be "more than 10000 [...]", "approximately 10+ thousand [...]" is too convoluted)

```html
<span class="ds badge">
  <span aria-hidden="true">1k</span>
  <span class="u-off-screen">approximately 1000 items</span>
</span>
```

To made things more difficult, we [can't](https://dequeuniversity.com/rules/axe/4.4/aria-hidden-focus) use `abbr` inside the `aria-hidden` container so it works for sighted users 🫠. (I can see why, for instance, YouTube ended up _not_ explaining "m" stands for million in the visual tooltip haha)

---

Another thought. Red Hat mentions in the [accessibility documentation](https://ux.redhat.com/elements/badge/accessibility/) for their badge that:

> Ensure that surrounding content can convey the purpose of a badge via assistive technologies

Given that badges are non-focusable (they're not interactive), it is likely that screen readers got there through a logical sequence instead of tabbing, so user is most likely aware of context. Let's keep it in mind (that's why I think we can do without "exist" in "[...] items exist").

**Comment by dgtlntv** [comment]
Design system contribution meeting:
- We think the badge should only have numerical values for now. We might revisit ranges / less than if needed.
- We think explaining abbreviations (2k, 4T+ etc.) are not necessary (neither screen readers, nor in the actual UI)
- It might be good to add context to the badge by indicating what the badge is counting (eg. item)

**Comment by jmuzina** [comment]
>   It might be good to add context to the badge by indicating what the badge is counting (eg. item)

@dgtlntv  just to clarify - do you mean adding count type (like "item") to the aria label or to the badge text?
